Encrypted Client Hello support (client only)

cpu opened the pull request:
## Description

This branch brings encrypted client hello (ECH) compatibility to Rustls **clients** based on [draft-ietf-tls-esni-18](https://datatracker.ietf.org/doc/html/draft-ietf-tls-esni-18). 

Closes https://github.com/rustls/rustls/issues/199

## Limitations

* It does _not_ add any support for writing Rustls' servers that accept ECH (in either shared or split mode). I think client support is the most important at this stage, and this branch is already very large/complex. We will track server-side support in https://github.com/rustls/rustls/issues/1980

* Extension compression is implemented separately in https://github.com/cpu/rustls/pull/6

<details>
<summary>Comparison to previous PRs:</summary>

## Server Name Enum

Initially we expected to update the `ServerName` enum to have a variant for ECH - I tried this initially and found that it didn't fit the design of ECH very well. Mainly the issue is that we need to carry a lot of additional state beyond the inner SNI name. Putting that state into the `ServerName` is challenging, and an impedance mismatch with the existing usage. Having separate state _and_ the ECH variant in `ServerName` meant having two pieces of information and having to contend with what to do if one is present but not the other. This is also made increasingly complex since we lifted the `ServerName` type into the pki-types crate. As a point of reference, you can compare this branch with the approach in https://github.com/rustls/rustls/pull/663 where a `ServerIdentity` enum stands in for `ServerName` and contains a `EncryptedClientHello(Box<EncryptedClientHello>)` variant that holds a `Box` of heap allocated state.

## Credit where credit is due

I started this branch by first adopting work from what @sayrer started in https://github.com/rustls/rustls/pull/663. Many thanks for kicking this effort off and giving me a solid foundation to start with!

Some major changes compared to that branch:

* This branch updates from draft-10 to draft-18, rolling in associated changes (mostly related to confirmation, extension types, etc).
* As discussed above, I opted to not use a `ServerName` variant for holding the inner SNI name and associated state.
* I've caught up the diff to the current Rustls `main`, adjusting for various `Codec` changes and associated code drift.
* HPKE usage is handled through the trait, abstracting away the choice of implementation as opposed to using hpke-rs directly.
* An HPKE provider backed by aws-lc-rs was implemented so we can have a crypto provider out-of-box that supports ECH.
* I've implemented support for ECH confirmation across Hello Retry Requests - this in turn required reworking how the "inner" transcript is maintained, calculating the HRR confirmation, and threading through more bits of state across the handshake.
* I've implemented support for session resumption in the inner hello when using ECH, including support for the GREASE PSK in the outer hello.
* I've tried to avoid duplicating encoding code, instead introducing an encoding `Purpose` that lets us share the message encoding logic as much as possible with the exception of the parts that are unique for ECH confirmation.
* I've tried to roll requirements into the existing code where possible to avoid duplication (e.g. updating the hash transcripts to allow forking, updating the key derivation code to allow confirmation calculation).
* I've wired in the "ECH required" alert, and pulling out ECH retry configs from the server's encrypted extensions. The retry configs and alert are held until the end of the handshake per spec.
* The ECH config HPKE suite is considered w.r.t FIPS status of the client config, or client connection, when using ECH.
* I've implemented test coverage with boringssl's bogo test suite, fixing divergences as appropriate.
* Probably some additional items I'm forgetting!

</details>

ctz commented:
> As an intermediate step it might be worthwhile to add a Go/OpenSSL server tests in CI (input here welcome!).

Definitely worth looking at the boringssl test suite -- there are ECH tests there that should be ready to go (albeit with some neccessary additions to our bogo_shim.rs). Currently they are disabled here:

https://github.com/rustls/rustls/blob/main/bogo/config.json#L34

cpu commented:
> Definitely worth looking at the boringssl test suite -- there are ECH tests there that should be ready to go

Oh great! Thanks for the pointer. I will focus attention there :+1:

djc commented:
Might also be useful to do a bogo update before this work? I assume ECH tests may have evolved since June.

djc commented on rustls/src/client/client_conn.rs:
```diff
@@ -386,6 +389,93 @@ pub enum Tls12Resumption {
+    // TODO(@cpu): update docstring with a pointer to the example program using hickory-dns.
+    // TODO(@cpu): consider dedicated HpkeError type(s).
+    pub fn new(
+        ech_configs: impl AsRef<[u8]>,
```
Hmm, I don't love the very generic parameter type here. Here's where it might be interesting to have rustls-pki-types type that represents a `Vec<EchConfigMsg>` somehow such that hickory could yield that type? (Maybe too ambitious for this stage.)

djc commented on rustls/src/client/client_conn.rs:
```diff
@@ -386,6 +389,93 @@ pub enum Tls12Resumption {
+
+                None
+            })
+            .ok_or_else(|| Error::General("No compatible ECH config".into()))?;
```
Style nit: I think this might be clearer as a `for`-loop or a `let (config, suite) = loop { .. }`.

cpu replied:
I pulled this out into a separate fn that uses a for loop.

djc commented on rustls/src/client/client_conn.rs:
```diff
@@ -255,6 +262,24 @@ impl ClientConfig {
+    ///
+    /// If the client configuration has enabled TLS 1.2, this function will return an error. ECH
+    /// may only be used with TLS 1.3+.
+    pub fn enable_ech(&mut self, config: EchConfig) -> Result<(), Error> {
```
So this is a bit odd in the sense that so far, we've built `ClientConfig`s by either going through the `ConfigBuilder` or just setting `pub` fields, and this is doing neither. I was thinking about whether it's possible to make the 1.2 xor ECH choice a compile-time failure but that probably involves adding another config builder state which we may not want?

Maybe we could provide a `with_ech()` method in in the `ConfigBuilder<S, WantsVersions>` state, which forces 1.3 only?

cpu replied:
I've tried a `with_ech` approach. I think it's better but there's a couple rough spots left:

1. We don't want to define this on `ConfigBuilder<S, WantsVersions>`, but `ConfigBuilder<ClientConfig, WantsVersions>` - the server and client configuration for ECH are quite different (e.g. a server might have multiple lists of ECH config, and will want to select a subset for offering as retry configs to facilitate graceful turn-down of old configs).

2. As a result, we need to stash a bit of client-specific state from the `impl ConfigBuilder<ClientConfig, WantsVersions> { fn with_ech(...) ... }` fn. That's not something supported by the builder architecture in general, only the last step of the build process can put things into the `ClientConfig`, and the `WantsVersion` state where we normally stash in-progress-build state is shared for both the client/server. The end result is the state gets an `Option` field that's specific to clients and ignored by builders. A bit gross, but I think workable since it's an internal detail.

3. There's _still_ an incompatible config condition we can't detect until we try to actually do ECH: the `ClientConfig` struct has a `enable_sni` `bool` field that could be set to `false` _after_ the config is built with ECH support. Configuring ECH but not sending SNI doesn't make sense, and this manifests as a new `EncryptedClientHelloError::SniRequired` error being returned from `hs.rs`'s `start_handshake` when ECH was configured but SNI disabled.

I don't think the above are deal-breakers, just wanted to call out that there's still some room for improvement.

djc commented on rustls/src/client/ech.rs:
```diff
@@ -0,0 +1,509 @@
+    ProtocolVersion, Tls13CipherSuite,
+};
+
+/// An enum representing ECH offer status.
+pub(crate) enum Status {
+    /// ECH was not offered - it is a normal TLS handshake.
+    NotOffered,
+    /// ECH was offered and the server accepted.
+    Accepted,
+    /// ECH was offered and the server rejected.
+    Rejected,
+}
+
+/// Randomly generated X25519 public key to use for GREASE ECH HPKE.
+const DUMMY_25519_PUBKEY: &[u8] = &[
+    0x67, 0x35, 0xCA, 0x50, 0x21, 0xFC, 0x4F, 0xE6, 0x29, 0x3B, 0x31, 0x2C, 0xB5, 0xE0, 0x97, 0xD8,
+    0xD0, 0x58, 0x97, 0xCF, 0x5C, 0x15, 0x12, 0x79, 0x4B, 0xEF, 0x1D, 0x98, 0x52, 0x74, 0xDC, 0x5E,
+];
```
Style nit: these should go to the bottom.

cpu replied:
Fixed.

djc commented on rustls/src/client/ech.rs:
```diff
@@ -0,0 +1,509 @@
+}
+
+pub(crate) fn fatal_alert_required(
+    common: &mut CommonState,
+    retry_configs: Option<Vec<EchConfigMsg>>,
```
Style nit: invert argument order?

cpu replied:
Done.

djc commented on rustls/src/client/ech.rs:
```diff
@@ -0,0 +1,509 @@
+    pub(crate) fn ech_hello(
+        &mut self,
+        mut outer_hello: ClientHelloPayload,
+        retryreq: Option<&HelloRetryRequest>,
```
Nit: `retry_req`?

cpu replied:
Renamed.

djc commented on rustls/src/client/ech.rs:
```diff
@@ -0,0 +1,509 @@
+        match ConstantTimeEq::ct_eq(derived.as_ref(), server_hello.random.0[24..].as_ref()).into() {
+            true => {
+                trace!("ECH accepted by server");
+                Ok(Some((inner_transcript, self.inner_hello_random)))
```
Style nit: hoist the `Ok`?

cpu replied:
Done.

djc commented on rustls/src/client/ech.rs:
```diff
@@ -0,0 +1,509 @@
+            // If none is found, the server has rejected ECH and there are no retry
+            // configs to offer. Abort with an ech required alert and return an err.
+            None => {
+                return Err(fatal_alert_required(common, None));
```
Style nit: this can sometimes be more compact if you omit the `;` (because it doesn't need a block).

cpu replied:
Good call, done.

djc commented on rustls/src/client/ech.rs:
```diff
@@ -0,0 +1,509 @@
+}
+
+impl EchState {
+    pub(crate) fn new(
```
Style nit: reconsider method order? Seems some simpler methods could drift towards the bottom.

cpu replied:
I reordered slightly, but I think for the most part this is in the preferred order already.

djc commented on rustls/src/client/hs.rs:
```diff
@@ -142,6 +143,18 @@ pub(super) fn start_handshake(
 
     let random = Random::new(config.provider.secure_random)?;
 
+    let ech_context = match &config.ech_config {
```
Style: `as_ref().map()`?

cpu replied:
I switched to `as_ref`, but using `map` isn't ergonomic here because I want to `?` the `new()` error into an early return and can't do that from within the `map` closure.

djc commented on rustls/src/client/hs.rs:
```diff
@@ -289,16 +305,45 @@ fn emit_client_hello_for_retry(
+    match (ech_context.as_mut(), config.grease_ech_hpke_provider) {
+        // ECH config, GREASE is irrelevant.
+        (Some(ech_context), _) => {
+            let ech_chp_payload = ech_context.ech_hello(chp_payload, retryreq)?;
```
Nit: inline the `ech_chp_payload` binding?

cpu replied:
Done.

djc commented on rustls/src/client/hs.rs:
```diff
@@ -289,16 +305,45 @@ fn emit_client_hello_for_retry(
+        }
+        // No ECH config, GREASE provider is present.
+        (None, Some(hpke_provider)) => {
+            let grease_ext = EchState::grease_ech_ext(
```
Same here.

cpu replied:
Done.

djc commented on rustls/src/client/hs.rs:
```diff
@@ -289,16 +305,45 @@ fn emit_client_hello_for_retry(
-            compression_methods: vec![Compression::Null],
-            extensions: exts,
-        }),
+        payload: HandshakePayload::ClientHello(chp_payload),
```
Nit: maybe do a separate commit that reorders the `ClientHelloPayload` instantiation and `sent_extensions` before adding the ECH stuff?

cpu replied:
Done.

djc commented on rustls/src/client/tls13.rs:
```diff
@@ -162,6 +165,30 @@ pub(super) fn handle_server_hello(
+                server_hello,
+                suite.common.hash_provider,
+            )? {
+                // The server rejected our ECH offer.
+                None => ech::Status::Rejected,
```
Style nit: I'd move this boring case down.

cpu replied:
Done.

djc commented on rustls/src/client/tls13.rs:
```diff
@@ -392,6 +421,18 @@ impl State<ClientConnectionData> for ExpectEncryptedExtensions {
+            let retry_configs = exts
+                .get_server_ech_extension()
+                .map(|ext| ext.retry_configs);
+            return Err(ech::fatal_alert_required(cx.common, retry_configs));
```
Could `ech::fatal_alert_required()` just take `&exts` instead?

cpu replied:
No, I don't think that will work well because we're currently using `fatal_alert_required` in two places:

1. From `confirm_hrr_acceptance` via `ExpectServerHelloOrHelloRetryRequest` when we discover ECH wasn't accepted during a hello retry request.
2. From `handle` in `State<ClientConnectionData> for ExpectEncryptedExtensions`.

In the second place we have a `exts` to pass through, but in the first place we don't have any server extensions yet and so use `None`.

djc replied:
Fair. Maybe consider just inlining the binding.

cpu replied:
Done.

djc commented on provider-example/examples/ech-client.rs:
```diff
@@ -0,0 +1,98 @@
+}
+
+// TODO(@cpu): share code with connect test, return all configs not just first.
+fn lookup_ech(resolver: &Resolver, domain: &str) -> Vec<u8> {
+    resolver
+        .lookup(domain, RecordType::HTTPS)
+        .expect("failed to lookup HTTPS record type")
+        .record_iter()
+        .find_map(|r| match r.data() {
+            Some(RData::HTTPS(svcb)) => svcb
+                .svc_params()
+                .iter()
+                .find_map(|sp| match sp {
+                    (SvcParamKey::EchConfig, SvcParamValue::EchConfig(e)) => Some(e.clone().0),
+                    _ => None,
+                }),
+            _ => None,
+        })
+        .expect("missing expected HTTPS SvcParam EchConfig record")
+}
```
Maybe we should upstream this into Hickory?

cpu replied:
You would have a better sense than I would about whether this would be something that would be accepted :-) It feels a little bit special purpose, but I can also see other folks having to write this same code so :shrug:

djc replied:
I don't think this is (much) more special purpose than things like [`tlsa_lookup()`](https://docs.rs/hickory-resolver/latest/hickory_resolver/struct.AsyncResolver.html#method.tlsa_lookup). Worth a discussion with Benjamin, at least.

cpu replied:
I left a TODO to think about this. I think it's best considered as an entirely separate follow-up since Hickory is lagging on Rustls versions and it will be a while before something upstreamed would be usable here.

djc reviewed (commented):
Looking good! I have a bunch of nits but not much substantial feedback.

I feel like the commit history overindexes a bit on small commits and might benefit from some squashing of things together so more context is available in commits.

cpu commented on rustls/src/enums.rs:
```diff
@@ -41,7 +41,8 @@ enum_builder! {
         CertificateRequired => 0x74,
-        NoApplicationProtocol => 0x78
+        NoApplicationProtocol => 0x78,
+        EncryptedClientHelloRequired => 0x79
```
I think it will require an update to the macro definition. I recall hitting errors with one. I'll see what I can come up with.

cpu replied:
https://github.com/rustls/rustls/pull/1725

cpu commented on rustls/src/msgs/handshake.rs:
```diff
@@ -2455,11 +2481,53 @@ impl Codec for EchConfigContents {
+}
+
+impl EchConfigExtension {
+    pub(crate) fn get_type(&self) -> ExtensionType {
```
I'm a bit torn on this feedback. The rest of `handshake.rs` uses `get_` prefixes. Maybe I should do a separate PR that tidies that up?

djc replied:
Yeah, might make sense to clean this up separate (also happy to suspend this review comment if you want to get this merged first).

cpu replied:
Happy to handle separately. I think I'll have to do a fair amount of rebasing and other work for this branch anyway. It's not something I expect to get merged ASAP.

djc replied:
Might also make sense to split this by getting all the ordering, derive twiddling merged early in a separate PR, which will prevent bitrot in those parts and make it easier to focus on the meat of ECH in this PR.

cpu replied:
That makes sense to me :+1:

cpu replied:
> Naming nit: maybe don't proliferate the use of get_() prefixes?

https://github.com/rustls/rustls/pull/1726

cpu replied:
> Might also make sense to split this by getting all the ordering, derive twiddling merged early in a separate PR, which will prevent bitrot in those parts and make it easier to focus on the meat of ECH in this PR.

I think I've pulled out all the easy stuff now :-)

djc replied:
Yeah, that's very helpful! I think it will make getting this merged much easier.

cpu commented on rustls/src/crypto/hpke.rs:
```diff
@@ -53,6 +53,16 @@ pub trait Hpke: Debug + Send + Sync {
+    ///
+    /// Returns both an encapsulated ciphertext and a sender context that can be used to seal
+    /// messages to the recipient.
+    fn setup_sender(
```
>  I'd invert the order (maybe a docstring could be useful here?).

Makes sense to me.

> Also, is pk_r a name used in the standard? Seems unidiomatically terse (again, a docstring could help).

Yeah, RFC 9180 and draft-ietf-tls-esni-17 both use "pkR" throughout. I changed that to `pk_r` to match Rust convention, but we could switch to a more idiomatic arg name. I don't think we need to match the standard here.

djc replied:
I would probably just call it `pub_key` and reference `pkR` in the doc string.

cpu replied:
Fixed up in https://github.com/rustls/rustls/pull/1737

cpu commented on rustls/src/crypto/hpke.rs:
```diff
@@ -66,6 +76,36 @@ pub trait Hpke: Debug + Send + Sync {
+///
+/// This is a stateful object that can be used to seal messages for receipt by
+/// a receiver.
+pub trait HpkeSender: Debug + Send + Sync {
```
RFC 9180 uses "Sender" and "Recipient" and then terse names like "SetupBaseS" and "SetupBaseR" for setting up the base mode sender and receiver, but it also uses "ContextS.Seal" and "ContextR.Open" for sealing/opening messages with an established sender/receiver context.

So in sum: I think using `Sealer` and `Opener` would be OK here unless someone has a stronger opinion :+1:

cpu replied:
Renamed in https://github.com/rustls/rustls/pull/1737

cpu commented on rustls/src/msgs/handshake.rs:
```diff
@@ -1182,6 +1177,54 @@ impl HelloRetryRequest {
+/// In some cases a handshake message may be encoded differently depending on the purpose
+/// the encoded message is being used for. For example, a [ServerHelloPayload] may be encoded
+/// with the last 8 bytes of the random zeroed out when being encoded for ECH confirmation.
+pub(crate) enum Purpose {
```
> Naming nit: not sure I love the Purpose name here... HandshakeMode/HandshakeKind/Context? 

I'm not deeply attached to it :-) `Context` seems reasonable to me. I'm less keen on the "Handshake" variants because we use both types of encoding in the context of a handshake where ECH has been offered and I think this is more of a per-message idea.

>  Does the standard use a word for distinguishing between non-ECH and ECH handshakes?

I don't think so. E.g. in [the section](https://datatracker.ietf.org/doc/html/draft-ietf-tls-esni-17#backend-server) relevant for the server hello accept confirmation it describes "computing" the ServerHello. Ditto for the HRR version in 7.2.1 "The backend server begins by computing HelloRetryRequest as usual, except  ..."

djc replied:
The downside of `Context` is that it is very generic and also used for example in the type passed to `Future::poll()`.

cpu replied:
> The downside of Context is that it is very generic

That's true. Maybe `Encoding::Standard` and `Encoding::EchConfirmation`?

djc replied:
Yes, that seems better.

cpu replied:
Fixed.

cpu commented on rustls/src/hash_hs.rs:
```diff
@@ -58,6 +58,16 @@ impl HandshakeHashBuffer {
+    /// Fork the hash computation, producing another context that has the
+    /// same prefix as this one.
+    #[allow(dead_code)] // TODO(@cpu): remove in subsequent commit.
+    pub(crate) fn get_fork(&self) -> Self {
```
> Nit: get_() prefixes (also for HandshakeHash) -- maybe just fork() or forked()

Good call, we already use `fork` and `fork_finish` in the `crypto/hash.rs` interface so I think we should match here.

>  is there a reason to not just use Clone?

We could do that for `HandshakeHashBuffer` used before a hash algorithm is decided where the struct just holds `Vec<u8>` and `bool`, but we can't for `HandshakeHash` where we hold a `&'static dyn hash::Hash` that we need to call `.fork()` on. I think consistency between `HandshakeHashBuffer` and `HandshakeHash` is a good reason to avoid using `clone()` for one and `fork()` for the other

djc replied:
Hmm, I don't know. How is `fork()` really different conceptually from `clone()`? Could do a custom `Clone` impl that uses `fork()` to make it work.

cpu replied:
>  Could do a custom Clone impl that uses fork() to make it work.

That's a good point. I like that idea :+1:

cpu replied:
Fixed up in https://github.com/rustls/rustls/pull/1739

cpu commented on rustls/src/client/client_conn.rs:
```diff
@@ -386,6 +389,93 @@ pub enum Tls12Resumption {
+    /// Construct an EchConfig by selecting a ECH config from the provided bytes that is compatible
+    /// with the given HPKE provider.
+    ///
+    /// The config bytes should be sourced from a DNS-over-HTTPS lookup resolving the `HTTPS` service
+    /// record record type for the host name of the server you wish to connect to using ECH,
```
I might need to brush up on my DNS terminology here. I think rfc9460 describes it as an "HTTPS resource record" that is a "SVCB-compatible RR type". I'll adjust for readability

cpu replied:
I reworded this comment.

cpu commented on rustls/src/client/client_conn.rs:
```diff
@@ -386,6 +389,93 @@ pub enum Tls12Resumption {
+    ) -> Result<Self, Error> {
+        // Try to unmarshal the ECH config bytes as a list of EchConfig structs, or a single EchConfig
+        // TODO(@cpu): verify if this is needed or if we should always assume a Vec...
+        let ech_configs = match Vec::<EchConfigMsg>::read(&mut Reader::init(ech_configs.as_ref())) {
```
Looks like no :+1: I'll fix.

cpu replied:
Fixed.

cpu commented on rustls/src/msgs/handshake.rs:
```diff
@@ -803,7 +803,7 @@ impl ServerExtension {
 }
 
-#[derive(Debug)]
+#[derive(Debug, Clone)]
```
Pulled out the derive changes into https://github.com/rustls/rustls/pull/1738 (and fixed the ordering)

cpu commented:
Now that a bunch of the smaller work has been pulled out and merged I will try to find time to rebase this and address the remaining feedback from djc's first review pass.

As a quick note: ISRG wants me to focus on finishing some of the remaining no-alloc/async work ahead of prioritizing ECH completion, so I'll be picking up work for this branch after making more progress on that front.

ctz commented:
> Might also be useful to do a bogo update before this work? I assume ECH tests may have evolved since June.

#1740

cpu commented on rustls/src/msgs/handshake.rs:
```diff
@@ -2434,7 +2434,33 @@ pub struct EchConfigContents {
+}
+
+#[allow(dead_code)] // TODO(@cpu): Remove in subsequent commits.
+impl EchConfigContents {
```
Commit message updated.

cpu commented:
I needed a break from my current tasks so I rebased this branch and addressed some of the low hanging review feedback. I haven't resolved all of it yet (so probably not ready for a second pass).

cpu commented on rustls/src/enums.rs:
```diff
@@ -565,6 +565,20 @@ enum_builder! {
     }
 }
 
+enum_builder! {
```
@djc I'm picking up this work again and wanted to check in. I think reworking the commits so that updates like this one are introduced alongside the changes that require them as you requested makes sense :+1: 

How do you feel about something like breaking up offering ECH and checking for ECH offer acceptance into two commits if it means putting a `todo!` or two in the first commit to stub out things that will be done in the second? It might be helpful to try and reduce the individual commit size but comes at the cost of making the overall "feature" not contained in a single commit. WDYT?

djc replied:
I'm open to it if you think it makes the PR easier to review overall.

cpu replied:
Thanks, that's useful to know. I'll see what I can come up with trying to balance commit size & avoiding too much churn.

cpu replied:
I've reworked based on the strategy I pitched above. I think it's the best balance of reviewability and keeping commits a manageable size. I think the way the protocol is implemented in a few pieces is also a very useful aide for others looking at implementing the RFC because it's _very_ non-linear compared to this commit stack.

sayrer commented:
It all looks good to me.

I think the Server Name Enum choices are good. I just threw stuff on the heap because I didn't really know what was going to be required. You're right that the spec is complicated and out-of-order when you sit down to write the implementation. I'd recommend reading it all the way through once before writing any more code, just to keep it all in your head. That's what I had to do a few times, anyway.

The HPKE trait is good, too. I wrote mine when rustls only had ring and didn't let one vary the dependencies, and was assuming ring would add that. I see you already have the OpenSSL server in your tests. You can find some variations that will trigger HRR via your HPKE choice here:
https://mailarchive.ietf.org/arch/msg/tls/DmMuf-CCzNNxSNY9LvxxQlu-fy4/

Those might be good for connect tests, and as a way to go once the server code is added.

cpu commented:
Hey sayrer,

> It all looks good to me.

Awesome. Thanks so much for taking a look. I'm focused on this branch again as of the past couple of days and hope we can get it across the line soon.

> I think the Server Name Enum choices are good. I just threw stuff on the heap because I didn't really know what was going to be required. 

Makes sense :+1: 

> You're right that the spec is complicated and out-of-order when you sit down to write the implementation. I'd recommend reading it all the way through once before writing any more code, just to keep it all in your head. That's what I had to do a few times, anyway.

Yeah :-( That's what I ended up doing too. It's some comfort to hear I wasn't alone in feeling this. From my perspective it feels like it's a bit too late in the document's process to try and fix the "editorial narrative" to be closer aligned to what an implementer might want. That's unfortunate but c'est la vie.

> You can find some variations that will trigger HRR via your HPKE choice here:
https://mailarchive.ietf.org/arch/msg/tls/DmMuf-CCzNNxSNY9LvxxQlu-fy4/

Useful pointers, thank you :bookmark: 

> As a practical matter, I think you will have to implement the handshake compression stuff. I showed up because I saw that you've added support for Kyber/X25519. In that case, you will really need it.

100% agreed. IIRC your initial branch had an implementation of the compression (or at least the start of one). I think we'll want to revive that before merging this work. I left it out only to try and keep the scope contained while figuring out the bigger picture.

Thanks again for the feedback (and the original work!).

sayrer commented:
> 100% agreed. IIRC your initial branch had an implementation of the compression (or at least the start of one). I think we'll want to revive that before merging this work. I left it out only to try and keep the scope contained while figuring out the bigger picture.

I think it was a little more ready for that part than this branch, but it didn't do it either. It's been a while, but it really wanted some Iter features that were nightly-only at the time. I think you can probably get a good version now.

sayrer commented:
> For that reason we may want to keep this as an experimental feature, and potentially hold merging the work?

I don't think you need to be so cautious. Firefox, Chrome, and Cloudflare are all shipping it.

<img width="767" alt="Screenshot 2024-03-28 at 9 48 49 AM" src="https://github.com/rustls/rustls/assets/253473/9d66d0c1-2360-4966-8311-0680e633f054">

dconnolly commented:
> > For that reason we may want to keep this as an experimental feature, and potentially hold merging the work?
> 
> I don't think you need to be so cautious. Firefox, Chrome, and Cloudflare are all shipping it.


ECH has gone through working group last call so it's unlikely to change significantly beyond version -18

cpu commented:
> cpu force-pushed the cpu-client-ech branch from 9e8d367 to 68cfb74 

Pushed some progress from my local branch:

* changed the user facing API for configuring ECH to fit into the builder paradigm.
* git patch on https://github.com/rustls/pki-types/pull/46 to experiment with a typed representation for the TLS encoded ECH config bytes. (_Note; this also temporarily breaks the cargo semver and fuzzer targets/clippy since they won't play nice with the git patch_).
* adjusted to be no-std compatible throughout, removing a pile of `#[cfg(feature = "std")]` gates I threw all over to get the build passing before.
* updated the ECH config handling to properly detect unsupported but mandatory ECH config extensions.
* reworked the GREASE implementation to provide more control to the user, and to fit better into the new builder model.
* reworked the commit history to prefer a style where pieces are introduced when they're used instead of ahead of time with `dead_code` allows.
* some other small bits of tidying/refactoring.

I'm still working through bogo coverage and that's now my major focus. I've grokked the overall test approach and done some bogo_shim updates and initial results indicate there's a ways to go w.r.t protocol compatibility in the cornercases. In particular the client ECH bogo tests are flagging that the inner hello contains some extensions it shouldn't, and most importantly, that I believe I've broken resumption in the presence of ECH. Hope to have some more updates on this front shortly.

cpu commented on rustls/src/client/ech.rs:
```diff
@@ -26,6 +27,32 @@ use crate::{
+    ///
+    /// This is an anti-ossification measure, sometimes referred to as "GREASE"[^0].
+    /// [^0]: <https://www.rfc-editor.org/rfc/rfc8701>
+    Placeholder(EchPlaceholderConfig),
```
Note: I've deliberately avoided using "GREASE" in the user-facing names for this. I think it's much too in-the-weeds to be a useful term for a public API. "Placeholder" seemed better, but I'm open to bikeshedding.

sayrer replied:
Totally doesn't matter. I know more about TLS than most, but when I saw "GREASE", I thought "oh what is this?" Maybe with some more colorful language.

I think you can get some good test material from the draft-07 code. You'll have to update it, but basically you add the "hex" crate as a dev dependency and then you can check every phase of the ECH code, but keep it in ASCII.

djc replied:
> Note: I've deliberately avoided using "GREASE" in the user-facing names for this. I think it's much too in-the-weeds to be a useful term for a public API. "Placeholder" seemed better, but I'm open to bikeshedding.

I feel like chances of `Grease` getting recognized as what it is are better than `Placeholder` getting recognized as the mechanism for grease? So I'd prefer to stick to the more familiar name.

sayrer replied:
Really doesn't matter, but GREASE is ok too. It's an acronym, so it should be all-caps:

https://www.rfc-editor.org/rfc/rfc8701.html

"Generate Random Extensions And Sustain Extensibility"

But note that had to capitalize "And", which one not normally do in English. :)

djc replied:
Rust naming conventions prescribe that acronyms get title-cased, so I think that spelling is idiomatic.

sayrer replied:
Yeah, I can see your point here. It's just that it is a wrong acronym, so its title-casing is also a wrong thing. But it's probably best to say "Grease" and never discuss it again.

cpu replied:
>  So I'd prefer to stick to the more familiar name.

I don't feel strongly, I will revert to using `Grease` :+1:

cpu commented:
> In particular the client ECH bogo tests are flagging that the inner hello contains some extensions it shouldn't, and most importantly, that I believe I've broken resumption in the presence of ECH. Hope to have some more updates on this front shortly.

The bogo tests uncovered a variety of issues (fun!) that I've now addressed. Some big changes to call-out:

* The config processing, pki-types, and hickory-dns interactions were all re-worked based on an updated understanding of the TLS encoding and some bugfixes I made in hickory (https://github.com/hickory-dns/hickory-dns/pull/2183) and the pki-types repo (https://github.com/rustls/pki-types/pull/46)
* Session resumption was totally busted. The binders weren't ever being filled in. Oops! This is fixed now, though I think I've still messed up the key schedule for early data w/ ECH acceptance.
* Related to the above, the outer hello now correctly GREASE-s its PSK extension when the inner hello is resuming. 
* Sent extensions are now tracked separately for the inner/outer hello. We need to do this so that if ECH is accepted we enforce no unsolicited extensions based on what was in the **inner** hello, since it may differ from the outer hello (e.g. inner could omit SNI, outer will always have SNI).
* ECH rejection is now handled at the end of the handshake instead of after encrypted extensions. The spec is clear on this, I just biffed it in my first go. When ECH is rejected we're supposed to continue handshaking with the outer hello, verifying the server certificate, etc. Notably we also need to refuse to send a client cert if requested when ECH was rejected. This is handled correctly now. Doing so requires carrying the potential retry configs through more of the client handshake states.
* The representation of ECH configurations within an ECH config list was changed so that we can treat configs with an unsupported version as opaque, skipping over them to consider other configs we may support.
* Lots of other small adjustments here and there, for example filtering out TLS1.2 compatibility extensions from the inner hello. 

My commit updating the bogo shim is a little bit rough. I recommend checking out the commit message for that update as it also outlines the updated test config. Top-level notes:

* We have to take a gross temporary dev-dep in Rustls on the provider-example crate to have an HPKE provider. Because the provider example uses hickory-dns we end up dragging in an older version of rustls through it as a transitive dependency and so have to start qualifying some commands in CI and the bogo runner. 
* I've been testing with ring (it's the bogo `runme` default provider) and discovered late that the aws-lc-rs HKDF code is panicing in the ECH confirmation code when running bogo with that provider. I will investigate/fix that ASAP.
* There's ~5 tests that are failing that I believe are indicative of problems with our impl. Need to chase these down.
* We ignore a handful of tests that aren't appropriate, or need small fixes upstream in bogo.
* I had to drag the bogo options struct deeper into the client/server processing in order to conditionally map errors based on config.

Phew! There's still a decent amount of work to do, but I feel more confident having gotten some decent bogo progress working. It continues to be difficult to impl this spec and I'm _very_ grateful bogo had test coverage to help catch some errors...

djc commented:
>     * We have to take a gross temporary dev-dep in Rustls on the provider-example crate to have an HPKE provider. Because the provider example uses hickory-dns we end up dragging in an older version of rustls through it as a transitive dependency and so have to start qualifying some commands in CI and the bogo runner.

It looks like Quinn 0.11 is pretty close, so we can hopefully get a fresh hickory release soon after.

ctz commented:
> * on the provider-example crate to have an HPKE provider.

AFAIK aws-lc-rs 1.6.0 added support needed for us to construct HPKE on top of it. I haven't gone back and looked at that again, but that might be a cleaner long-term arrangement.

cpu commented:
> that might be a cleaner long-term arrangement.

Agreed - my rough plan is to:

* Figure out the aws-lc-rs provider panic
* Figure out the ~5 remaining bogo tests I think require adjustment on our side
* Look at implementing the HPKE provider interface w/ aws-lc-rs's updated primitives. 

I'm expecting to land the aws-lc-rs HPKE provider in a separate PR from this one - it's a nice isolated chunk of work. My hope is we can drive reviews forward here in the meantime, land the HPKE provider, update this branch to use it, and then remove the dependency on the provider-example. Nothing in the ECH work should have to change since the HPKE usage is behind a trait.

At that point we can either not test ECH w/ `*ring*` (it doesn't expose the APIs we need), or have the `*ring*` bogo tests use `*ring*` for the crypto provider, but aws-lc-rs as the HPKE provider (that seems reasonable to me and less hacky than depending on the provider-example).

cpu commented:
> Figure out the aws-lc-rs provider panic

I traced this down to a limit that aws-lc-rs was placing on the maximum allowable `info` size for an HKDF expansion. The limit in use was 80 bytes, but ECH confirmation calculations using SHA-384 or SHA-512 for the transcript digest produces `info` parameters of 81 and 97 bytes, exceeding the limit. I've proposed a limit increase in https://github.com/aws/aws-lc-rs/pull/411 and applying a git patch on `aws-lc-rs` in this branch allowed restoring the `bogo` coverage for `aws-lc-rs` without any additional changes. `*ring*`'s impl has no similar limit and so was working without error already.

ctz commented on rustls/src/msgs/handshake.rs:
```diff
@@ -2532,39 +2552,109 @@ impl Codec<'_> for EchConfigContents {
-            contents: EchConfigContents::read(&mut r.sub(length as usize)?)?,
+
+        Ok(match version {
+            EchVersion::V18 => Self::V18(EchConfigContents::read(&mut r.sub(length as usize)?)?),
```
any value in extracting `r.sub(length as usize)?` -- seems like every match arm here must do that to be correct.

cpu replied:
Done.

ctz commented on rustls/src/client/ech.rs:
```diff
@@ -109,3 +136,364 @@ impl EchConfig {
+            _ => {
+                // L + 9
+                // "This is the length of a "server_name" extension with an L-byte name."
+                max_name_len + 9
```
I think `max_name_len` and `padding_len` are both u8 at this point (could be wrong) so `max_name_len + 9` could certainly overflow based on DNS  EchConfig contents. widen `padding_len` to usize?

cpu replied:
Very good catch, thank you. We were promoting `padding_len` to usize later on, so I shifted that to happen here to avoid the overflow and left a comment about it.

ctz reviewed (commented):
up to and including `ech: construct ECH client hello`

ctz commented on rustls/src/client/ech.rs:
```diff
@@ -299,6 +319,54 @@ impl EchState {
+        server_hello: &ServerHelloPayload,
+        hash: &'static dyn Hash,
+    ) -> Result<Option<EchAccepted>, Error> {
+        // Star the inner transcript hash now that we know the hash algorithm to use.
```
```suggestion
        // Start the inner transcript hash now that we know the hash algorithm to use.
```

cpu replied:
Fixed

ctz commented on rustls/tests/api.rs:
```diff
@@ -546,7 +546,7 @@ fn test_config_builders_debug() {
     assert_eq!("ConfigBuilder<ServerConfig, _> { state: WantsVersions { provider: CryptoProvider { cipher_suites: [TLS13_CHACHA20_POLY1305_SHA256], kx_groups: [X25519], signature_verification_algorithms: WebPkiSupportedAlgorithms { all: [ .. ], mapping: [ECDSA_NISTP384_SHA384, ECDSA_NISTP256_SHA256, ED25519, RSA_PSS_SHA512, RSA_PSS_SHA384, RSA_PSS_SHA256, RSA_PKCS1_SHA512, RSA_PKCS1_SHA384, RSA_P ...
     let b = server_config_builder_with_versions(&[&rustls::version::TLS13]);
     assert_eq!(
-        "ConfigBuilder<ServerConfig, _> { state: WantsVerifier { provider: CryptoProvider { cipher_suites: [TLS13_AES_256_GCM_SHA384, TLS13_AES_128_GCM_SHA256, TLS13_CHACHA20_POLY1305_SHA256, TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384, TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256, TLS_ECDHE_ECDSA_WITH_CHACHA20_POLY1305_SHA256, TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384, TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256,  ...
```
Feels like this test is actually more noisy than useful.

cpu replied:
> Feels like this test is actually more noisy than useful.

Yeah, agreed. I pulled out a change to make this test not assert the exact debug string content anymore and landed that before the ECH work.

ctz reviewed (commented):
Good stuff!

cpu commented on rustls/src/msgs/handshake.rs:
```diff
@@ -1072,7 +1072,7 @@ impl TlsListElement for HelloRetryExtension {
 }
 
-#[derive(Debug)]
+#[derive(Debug, Clone)]
```
Fixed the commit headline and removed the redundant sentence. Thanks for noticing

cpu commented on rustls/src/error.rs:
```diff
@@ -47,6 +47,9 @@ pub enum Error {
     DecryptError,
 
+    /// An error occurred while handling Encrypted Client Hello (ECH).
+    EncryptedClientHello(EncryptedClientHelloError),
```
Good call-out. I went with `InvalidEncryptedClientHello`

cpu commented on rustls/src/msgs/handshake.rs:
```diff
@@ -2511,7 +2511,27 @@ pub struct EchConfigContents {
+        self.extensions
+            .iter()
+            // An extension is considered mandatory if the high bit of its type is set.
+            .any(|ext| u16::from(ext.ext_type()) & 0x8000 != 0)
```
Ooo yes, good catch. I've added a `matches!` check.

cpu commented on rustls/src/msgs/handshake.rs:
```diff
@@ -2532,39 +2552,109 @@ impl Codec<'_> for EchConfigContents {
+                // Write the version, the length, and the contents.
+                EchVersion::V18.encode(bytes);
+                bytes.extend_from_slice(length);
+                bytes.extend(contents);
```
Ah! much cleaner. Implemented in commit.

cpu commented:
> Figure out the ~5 remaining bogo tests I think require adjustment on our side

I think I've gotten to the bottom of 4 of the 5 of these ignored tests (still TODO: `TLS-ECH-Client-EarlyData`). I'm going to have to do a little bit of refactoring to fix the issue (need to adjust HRR behaviour) so I'd suggest waiting before doing a review pass until I've pushed the update.

cpu commented:
> I'd suggest waiting before doing a review pass until I've pushed the update.

Done.

> I think I've gotten to the bottom of 4 of the 5 of these ignored tests

For the curious, for 3 of the 5 tests (`TLS-ECH-Client-Reject-HelloRetryRequest`, `TLS-ECH-Client-GREASE-IgnoreHRRExtension` and `TLS-ECH-GREASE-Client-TLS13-HelloRetryRequest`), it was sort of two issues. The first was rejecting too early in the handshake in the HRR case and the second was incorrectly regenerating the ECH or grease ECH ext when making a second hello in the case where HRR rejects ECH or we GREASE'd.
 
For 1 of the 5, `TLS-ECH-Client-Reject-ResumeInnerSession-TLS13`, I realized it expects an `:UNEXPECTED_EXTENSION:` error based on the BoringSSL ECH impl not sending [GREASE PSK](https://datatracker.ietf.org/doc/html/draft-ietf-tls-esni-18#section-6.1.2). This is a SHOULD that BoringSSL chose not to implement but we did. As a result we return  `:ECH_REJECTED:` for this test and have to ignore it. The resumption extension is only unexpected if the outer hello didn't include a PSK ext.

That means there's just one bogo failure left to chase down, `TLS-ECH-Client-EarlyData`. I will dig into this after some higher priority support work is settled.

cpu commented:
> That means there's just one bogo failure left to chase down, TLS-ECH-Client-EarlyData. I will dig into this after some higher priority support work is settled.

Fixed; I was using the wrong transcript/random in one more place related to the early data key schedule w/ ECH. All the bogo tests I think are relevant are now passing so I'm going to switch my focus to implementing an HPKE provider backed by aws-lc-rs.

cpu commented:
> cpu force-pushed the cpu-client-ech branch from 1fa3a32 to de52ea1 

* Dropped the aws-lc-rs patch since we're now using [1.7.1](https://github.com/aws/aws-lc-rs/releases/tag/v1.7.1) with the HKDF info maximum limit fix.
* Dropped the commits that were pulled out into https://github.com/rustls/rustls/pull/1944
* Rebased on `main`.

cpu commented:
> I'm going to switch my focus to implementing an HPKE provider backed by aws-lc-rs.

First pass for this up for review: https://github.com/rustls/rustls/pull/1963 I will rebase this branch on top and remove the provider-example workaround hacks shortly.

cpu commented:
In https://github.com/rustls/rustls/pull/1963 [Ctz asked](https://github.com/rustls/rustls/pull/1963#discussion_r1616140371):

> do we want to add a fips() function on ClientConnection/ServerConnection that delegates to the config version, but also takes into account of the crypto backing ECH?

I think the answer is yes - this is something we should impl in this branch once rebased on top of 1963 to have `Hpke::fips()`.

cpu commented:
How would we like to handle bogo testing w/ `*ring*` as the crypto provider w.r.t the ECH tests? I think there are two options:

1. We change the `bogo/runme` script to use rustls w/ both features (`ring` and `aws_lc_rs`) when `BOGO_SHIM_PROVIDER` is `ring`, and then use `*ring*` for everything as we do today, but use the `*aws-lc-rs*` HPKE suites for the ECH tests.
2. We make the `bogo_shim_impl.rs` smarter, and skip the ECH tests when the aws-lc-rs feature isn't enabled.

I don't have a strong preference here. Option 2 might be better for representing what rustls-with-only-ring is capable of, but will require some hacks in the shim.

djc commented:
> Now that #1963 is close to merge, how would we like to handle bogo testing w/ `*ring*` as the crypto provider w.r.t the ECH tests? I think there are two options:
> 
>     1. We change the `bogo/runme` script to use rustls w/ both features (`ring` and `aws_lc_rs`) when `BOGO_SHIM_PROVIDER` is `ring`, and then use `*ring*` for everything as we do today, but use the `*aws-lc-rs*` HPKE suites for the ECH tests.
> 
>     2. We make the `bogo_shim_impl.rs` smarter, and skip the ECH tests when the aws-lc-rs feature isn't enabled.
> 
> 
> I don't have a strong preference here. Option 2 might be better for representing what rustls-with-only-ring is capable of, but will require some hacks in the shim.

Option 2 sounds cleaner to me -- seems okay to have some hacks in the shim.

ctz commented:
> Now that #1963 is close to merge, how would we like to handle bogo testing w/ `*ring*` as the crypto provider w.r.t the ECH tests?

Can we do more of this sort of thing? https://github.com/rustls/rustls/blob/main/bogo/config.json.in#L68-L70

(nb. this is a list of _disabled_ tests, so in this case we are adding P521 tests to that only for the ring provider.)

cpu commented:
> cpu marked this pull request as ready for review now

At long last, I think this is ready for review in a non-draft sense. It's been rebased and updated to use the aws-lc-rs HPKE suites landed in https://github.com/rustls/rustls/pull/1963 The bogo suites are passing for both ring (no ECH tests) and aws-lc-rs (w/ ECH tests).

The one remaining piece of work is to implement extension compression. This is fairly tricky based on the extension randomization we landed in https://github.com/rustls/rustls/pull/1730. I'm going to work on that next, but since this is already a big branch I think it's reasonable to merge as-is and iterate on that aspect as a follow-up. ctz and bdaehlie both agreed when we discussed alongside roadmap planning.

ctz commented on rustls/src/client/hs.rs:
```diff
@@ -359,11 +387,24 @@ fn emit_client_hello_for_retry(
+    let early_key_schedule = match (ech_state.as_mut(), tls13_session) {
+        // If we're performing ECH and resuming, then the PSK binder will have been dealt with
+        // separately and we need to take the early_data_key_schedule computed for the inner hello.
+        // TODO(@cpu): This is an area I'm not 100% confident the implementation is correct!
```
is this something that has been addressed now, or is it still a worry? any ideas on how to gain some confidence?

cpu replied:
Oops, this slipped through from an earlier revision. I'm confident in the implementation now (and bogo concurs). I will remove the comment.

cpu replied:
Removed stale TODO.

ctz commented on examples/src/bin/ech-client.rs:
````diff
@@ -3,24 +3,63 @@
+//! ```
+//!
+//! This will perform a DNS-over-HTTPS lookup for the defo.ie ECH config, using it to determine
+//! the plaintext SNI ot send to the server. The protected encrypted SNI will be "www.defo.ie".
````
```suggestion
//! the plaintext SNI to send to the server. The protected encrypted SNI will be "www.defo.ie".
```

cpu replied:
Fixed.

ctz commented on examples/src/bin/ech-client.rs:
```diff
@@ -99,6 +217,29 @@ fn lookup_ech_configs(resolver: &Resolver, domain: &str) -> pki_types::EchConfig
+///
+/// A real implementation should vary this pubkey, ensuring the public key type matches the KEM
+/// selected in the GREASE [Hpke] instance.
+const GREASE_25519_PUBKEY: &[u8] = &[
```
Is it worth extending our Hpke traits to be able to generate fresh meaningless public keys, so we can remove this parameter from `EchGreaseConfig::new()`?

cpu replied:
That sounds reasonable to me. It might come in handy for https://github.com/rustls/rustls/issues/1980 as well.

cpu replied:
I'm still chewing on this. It's a little bit tricky, because we're using [`agreement::PrivateKey::from_bytes`](https://docs.rs/aws-lc-rs/latest/aws_lc_rs/agreement/struct.PrivateKey.html#method.from_private_key) to reconstitute a private key from the provided bytes (this is how the test vectors are represented), but the [trait impl](https://docs.rs/aws-lc-rs/latest/aws_lc_rs/agreement/struct.PrivateKey.html#trait-implementations) to go from a generated `PrivateKey` to raw BE bytes is different depending on the type of key: one supports only X25519 and one supports the P curves. I think I can get this working but wanted to push the other easier feedback first.

cpu replied:
Done. It's a separate commit up-front, and could be lifted into its own PR if desired.

ctz commented on rustls/src/client/client_conn.rs:
```diff
@@ -727,6 +730,15 @@ mod connection {
+        /// This is different from [`crate::crypto::CryptoProvider::fips()`]:
+        /// it is concerned only with cryptography, whereas this _also_ covers TLS-level
+        /// configuration that NIST recommends, as well as ECH HPKE suites if applicable.
+        pub fn fips(&self) -> bool {
```
Probably warrants a tests that calls it in api.rs.

cpu replied:
I added a test for the `ClientConfig::fips()` update. Still TODO: `ClientConnection::fips()` coverage.

cpu replied:
> Still TODO: ClientConnection::fips() coverage.

Added.

ctz commented on rustls/examples/internal/bogo_shim_impl.rs:
```diff
@@ -13,12 +13,22 @@ use base64::prelude::{Engine, BASE64_STANDARD};
 use rustls::client::{ClientConfig, ClientConnection, Resumption, WebPkiServerVerifier};
-#[cfg(all(not(feature = "ring"), feature = "aws_lc_rs"))]
+#[cfg(feature = "aws_lc_rs")]
+use rustls::client::{EchConfig, EchGreaseConfig, EchMode, EchStatus};
```
I don't mind either way, but I would write a big compound use here to avoid so many feature gates.

cpu commented on rustls/src/client/builder.rs:
```diff
@@ -6,14 +6,30 @@ use pki_types::{CertificateDer, PrivateKeyDer};
+    ///
+    /// This implicitly selects TLS 1.3 as the only supported protocol version to meet the
+    /// requirement to support ECH.
+    pub fn with_ech(
```
Updated with some guidance. I suspect we might want to tweak the wording further but it's a start!

cpu commented on rustls/src/client/ech.rs:
```diff
@@ -0,0 +1,108 @@
+    /// with one of the given HPKE suites.
+    ///
+    /// The config list bytes should be sourced from a DNS-over-HTTPS lookup resolving the `HTTPS`
+    /// resource record for the host name of the server you wish to connect to using ECH,
```
Applied suggested change.

cpu commented on rustls/src/client/ech.rs:
```diff
@@ -0,0 +1,108 @@
+
+            if contents.has_unknown_mandatory_extension() || contents.has_duplicate_extension() {
+                warn!(
+                    "ECH config {} has duplicate, or unknown mandatory extensions",
```
Done.

cpu commented on rustls/src/crypto/aws_lc_rs/hpke.rs:
```diff
@@ -896,66 +961,59 @@ mod tests {
+            let open_res = opener.open(&[0x0], &ct);
+            assert!(matches!(open_res.unwrap_err(), Error::Other(_)));
+
+            // Opening the plaintext with the wrong opener should fail.
+            let mut sk_rm_prime = sk.secret_bytes().to_vec();
+            sk_rm_prime[10] = 0x00; // Corrupt a byte of the private key somewhere after potential zero padding.
+            let mut opener_two = suite
+                .setup_opener(&enc, info, &HpkePrivateKey::from(sk_rm_prime))
+                .unwrap();
+            let open_res = opener_two.open(aad, &ct);
+            assert!(matches!(open_res.unwrap_err(), Error::Other(_)));
```
note to self; this test is flaky now that it's run for all suites with random keys instead of `DH_KEM_X25519_HKDF_SHA256_AES_128` w/ a fixed key - need to investigate why.

ctz replied:
I would suggest corrupting the key using `sk_rm_prime[10] ^= 0xff;` or similar -- just zeroing a byte in a uniformly random string will result in the same string 1 time out of 256.

ctz replied:
(assuming that is the issue!)

cpu replied:
>  just zeroing a byte in a uniformly random string will result in the same string 1 time out of 256.

:facepalm: Of course. That was silly of me. Fixed and no longer flaky in my local burst testing.

ctz commented on Cargo.toml:
```diff
@@ -28,3 +28,8 @@ resolver = "2"
+[patch.crates-io]
+# TODO(XXX): Remove this once 0.25 is released - we want the ECH fixes from
+#            https://github.com/hickory-dns/hickory-dns/pull/2183
+hickory-resolver = { git = "https://github.com/hickory-dns/hickory-dns", rev = "6334a01430088ead8642cafaee592ec7bf49831f" }
```
I _think_ the MSRV job is upset with this because this states a requirement of 1.67? This was fine without a [patch], because `connect-tests` is not part of the MSRV check job usually. But for whatever reason, cargo decides it needs to re-find this dependency to build something unrelated.

cpu replied:
I'm confused about what's going on with this CI task. Your theory about the `hickory-dns` MSRV seems plausible, but the output looks more like it can't resolve the patch content, not that the MSRV was incompatible.

I can reproduce locally w/ the 1.63 toolchain but haven't been able to figure out a fix yet.

cpu replied:
It does seem to work locally w/ 1.64, 1.65, 1.66, and 1.67.

Maybe there's a bug with patch resolution in 1.63?

cpu replied:
I wonder if it's related to 1.64+ having added [workspace inheritance](https://blog.rust-lang.org/2022/09/22/Rust-1.64.0.html#cargo-improvements-workspace-inheritance-and-multi-target-builds) since the hickory-dns project is using that: https://github.com/hickory-dns/hickory-dns/blob/f1489da675c21fddc189f2c9505bc9da6c156835/Cargo.toml#L29-L35

cpu replied:
> I wonder if it's related to 1.64+ having added workspace inheritance

I think that's the case unfortunately and I can't think of a great solution here :-/ Even though the patch isn't used in our MSRV build for the `rustls` lib (it's only used by the `connect-tests` and `examples` crates) we can't resolve the patch unless we're using 1.64+ and so the MSRV task is broken.

Some ideas:

* Wait for an upstream release of hickory-dns with the fixes we need - this seems crummy, we could be waiting for a while.
* Switch our MSRV to 1.64 - this seems crummy (and Debian stable is still on 1.63).
* Switch to using a hickory-dns release before the 1.64+ workspace inheritance change. This would mean backporting my SVCB fixes to trust-dns 0.20 and reverting the connect-test/example code to use the old API.

The last idea seems the least gross, but still not ideal :face_exhaling:

cpu replied:
> Wait for an upstream release of hickory-dns with the fixes we need - this seems crummy, we could be waiting for a while.

Ctz rightly points out we could make do with a published pre-release. I'll file an issue to inquire about this.

cpu replied:
> Switch to using a hickory-dns release before the 1.64+ workspace inheritance change. This would mean backporting my SVCB fixes to trust-dns 0.20 and reverting the connect-test/example code to use the old API.

While we're waiting I did a quick-n-dirty version of this (see `cpu/trust-dns` [cpu-0.22-ech-fixes branch](https://github.com/cpu/trust-dns/tree/cpu-0.22.0-ech-fixes) - I backported only the core ECH fixes, not the comment updates or the general SVCB improvements). The rustls side updates to use this compatible `patch` and the older APIs are in a separate commit at the end so that we can hopefully revert it once a pre-release is available.

Doing this uncovered one real MSRV issue with this branch: I was using a `let ... else` construction that's unstable on 1.63. I fixed this in the original commit that introduced it along with a TODO to revert to the nicer construction when we bump MSRV.

With both of the above in place this branch should be all green in CI.

cpu commented:
@djc Do you think you'll have a chance to review this branch this week?

djc commented:
Yup! Sorry for not getting to it sooner, I'll keep it at the top of my list for today/tomorrow.

cpu commented:
Thanks! Appreciated

cpu commented:
> The one remaining piece of work is to implement extension compression. 

I have this work staged in a [draft PR](https://github.com/cpu/rustls/pull/6) on my fork that targets this branch as its base.

djc commented on rustls/src/client/tls13.rs:
```diff
@@ -72,6 +73,8 @@ pub(super) fn handle_server_hello(
     hello: ClientHelloDetails,
     our_key_share: Box<dyn ActiveKeyExchange>,
     mut sent_tls13_fake_ccs: bool,
+    ech_state: Option<EchState>,
+    server_hello_msg: &Message,
```
Style nit: feels like `server_hello_msg` should come earlier in this list? At least before `ech_state`, which seems like it would be longer lived.

cpu replied:
Done.

djc commented on rustls/src/client/tls13.rs:
```diff
@@ -146,6 +149,10 @@ pub(super) fn handle_server_hello(
 
     let shared_secret = our_key_share.complete(&their_key_share.payload.0)?;
 
+    // TODO(@cpu): Handle confirmation of ECH.
```
As discussed a while ago, I question the value of leaving these loose ends in main branch history (relative to squashing more things here).

cpu replied:
I think this is the most awkward join point and it could be resolved by squashing the "ech: construct ECH client hello" commit with the subsequent "ech: confirm acceptance of ECH offer" commit.

I'll do that in the rebase that applies your feedback. :+1:

cpu replied:
> could be resolved by squashing the "ech: construct ECH client hello" commit with the subsequent "ech: confirm acceptance of ECH offer" commit.

I think we'll actually end up wanting three commits joined: the config setup, the hello construct, and the acceptance confirm.

cpu replied:
I ended up squashing the majority of the implementation into one commit that:

* adds the config elements
* generates the hello
* confirms acceptance
* confirms acceptance w/ HRR

The GREASE commit remains separate since it didn't require any intermediate TODOs.

djc commented on rustls/src/client/ech.rs:
```diff
@@ -292,6 +312,54 @@ impl EchState {
+        // Check that first 8 digits of the derived secret match the last 8 digits of the original
+        // server random. This match signals that the server accepted the ECH offer.
+        // Indexing safety: Random is [0; 32] by construction.
+        Ok(
```
Style nit: I think I've come to the conclusion that hoisting the `Ok` might not be worth it if it causes 2 extra lines + an extra level of indentation.

cpu replied:
Fixed.

djc commented on rustls/src/client/ech.rs:
```diff
@@ -493,4 +561,42 @@ impl EchState {
         Ok(())
     }
+
+    fn server_hello_conf(server_hello: &ServerHelloPayload) -> Message {
```
Maybe take an owned `ServerHelloPayload` if that's actually what we need, on the off chance that the caller doesn't need to keep their copy around?

cpu replied:
This ends up being tricky. We get our `&ServerHelloPayload` from `ExpectServerHello::handle` where presently we use `require_handshake_msg!` to extract the `&ServerHelloPayload` from `m: Message<'m>`. We read a `SupportedCipherSuite` from that `&ServerHelloPayload` and then use `m` again to start the outer hello handshake transcript with the `suite`.

If we switch to using `require_handshake_msg_move!` to get a `ServerHelloPayload` then `m` is partially moved and we can't use it to update the transcript anymore. We can't start the transcript before the move because we don't know the suite until we have a `ServerHelloPayload` to look at.

We also want the `&Message` for the new argument to `handle_server_hello` for updating the inner transcript after ECH has been confirmed, and it's a partially moved value if we pull out an owned `ServerHelloPayload` earlier.

It might be workable, but I'm pretty convinced that it will take a decent sized refactor I'm not super keen on starting at this stage. It feels like the callers always want to keep the `Message` around, payload intact.

djc replied:
Ahhh, no, wasn't aware of all that complexity hanging off of that. Definitely sounds like that isn't worth it.

djc commented on rustls/src/msgs/handshake.rs:
```diff
@@ -1198,16 +1204,7 @@ pub struct ServerHelloPayload {
 
 impl Codec<'_> for ServerHelloPayload {
     fn encode(&self, bytes: &mut Vec<u8>) {
-        self.legacy_version.encode(bytes);
```
On the other hand, this is something I might do in a separate commit.

cpu replied:
I think I'll end up with dead code warnings if I try to do that before implementing confirmation, or I'd have to implement a watered down version that only adds `Encoding::Standard`. In the latter case having that separated seems like it has less value. Do you feel strongly about this one?

djc replied:
Nope!

djc commented on rustls/src/client/ech.rs:
```diff
@@ -360,6 +361,74 @@ impl EchState {
+            confirmation_transcript.current_hash(),
+        );
+
+        Ok(
```
(Same thing for the `Ok`-wrapping here.)

cpu replied:
Fixed.

djc commented on rustls/src/client/ech.rs:
```diff
@@ -569,6 +638,13 @@ impl EchState {
+    fn hello_retry_request_conf(retry_req: &HelloRetryRequest) -> Message {
+        Self::ech_conf_message(HandshakeMessagePayload {
+            typ: HandshakeType::HelloRetryRequest,
+            payload: HandshakePayload::HelloRetryRequest(retry_req.clone()),
```
Take the owned `HelloRetryRequest`?

djc commented on rustls/src/msgs/handshake.rs:
```diff
@@ -1190,6 +1192,48 @@ impl HelloRetryRequest {
+            // See draft-ietf-tls-esni-18 7.2.1:
+            // <https://datatracker.ietf.org/doc/html/draft-ietf-tls-esni-18#name-sending-helloretryrequest-2>
+            Encoding::EchConfirmation => {
+                let extensions = LengthPrefixedBuffer::new(ListLength::U16, bytes);
```
Style nit: maybe take this out of the match arm?

cpu replied:
I thought that would be workable but it ends up creating two mutable borrows of `bytes` to lift the `LengthPrefixedBuffer` up. The `Standard` arm takes a mut borrow of `bytes` for `self.extensions.encode(bytes)`.

djc commented on rustls/src/client/hs.rs:
```diff
@@ -398,6 +397,25 @@ fn emit_client_hello_for_retry(
+        // If we haven't offered ECH, and have no ECH state, then consider whether to use GREASE
+        // ECH.
+        (EchStatus::NotOffered, None) => {
+            if let Some(EchMode::Grease(grease_config)) = &config.ech_mode {
```
Take this out of the match arm?

cpu replied:
Done.

djc reviewed (approved):
Okay, most of the core logic is over my head -- I don't have enough time/energy to fully digest it. But it looks well-designed, and I only have a few stylistic suggestions.

cpu commented on rustls/src/client/ech.rs:
```diff
@@ -0,0 +1,105 @@
+#[cfg(feature = "logging")]
+use crate::log::{debug, warn};
+use crate::msgs::codec::{Codec, Reader};
+use crate::msgs::handshake::EchConfig as EchConfigMsg;
```
I don't think there's a strong reason. I was trying to match to the names from [draft-ietf-tls-esni-18 §4](https://datatracker.ietf.org/doc/html/draft-ietf-tls-esni-18#section-4), but I think the `enum` arrangement already makes this not 1:1 with the simple presentation language in the spec.

Calling it `EchConfigPayload` sounds reasonable to me :+1: I will add a rename commit up-front and remove the aliased `use`'s.

cpu replied:
Done.

cpu commented on rustls/src/client/ech.rs:
```diff
@@ -0,0 +1,105 @@
+        Ok(Self { config, suite })
+    }
+
+    fn select_config_and_suite(
```
Good point; I will inline.

cpu replied:
Done.

cpu commented on rustls/src/client/hs.rs:
```diff
@@ -146,16 +147,18 @@ pub(super) fn start_handshake(
-            let _config = &ech_config.config;
-            Some(())
-        });
+    let ech_state = match config.ech_config.as_ref() {
```
Hmm, I think it might make things more awkward? I can't use an `if let / else` to match the `&EchMode` given to the `map` closure because of MSRV, so I'd still need a `match { }` in the `map` body making it about the ~same amount of code, but more annoying to deal with the fallibility of the `EchState::new()`.

Am I overthinking what you had in mind?

djc replied:
I missed the `?` operator. I can see how that makes this probably not worth it.

cpu commented:
After discussing with the other maintainers we're going to merge this with the dev-dependency patch in 36097a79f4705ff74f009cc6443e0cf2b4a3625b in-place, with the plan to revert that commit & the patch when there's a pre-release of hickory-dns 0.25
